Add Advisor Configurations API and api-version 2017-04-19

davidwa-msft opened the pull request:
Add Advisor Configurations API to api-version 2017-04-19.
Add Create, Read/List and Update APIs for new Advisor Configuration entity at Subscription and Resource Group scopes.

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [X] The title of the PR is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [X] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [X] If applicable, the PR references the bug/issue that it fixes.
- [X] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [X] My spec meets the review criteria:
  - [X] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [X] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [X] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

Commit:
Move swagger  file to swagger directory

Commit:
Update Advisor swagger for Configurations

marstr commented on specification/advisor/resource-manager/Microsoft.Advisor/2017-04-19/swagger/advisor.json:
```diff
@@ -269,65 +610,33 @@
 						"description": "The number of suppressions per page if a paged version of this API is being used.",
 						"required": false,
-						"type": "integer",
-						"format": "int32"
```
Keep in mind, for some languages this will cause breaking changes to their SDK. In Go for example, there isn't any implicit conversion so [one cannot seamlessly pass an `int` to an `int32`](https://play.golang.org/p/rXqYpY8Qy7) or vice-versa.

davidwa-msft replied:
Should we just pick a consistent type and go with it? For example, I see there is another "integer" exposed on this same swagger spec but it is "integer" instead. I would like to either consistently name our property types/formats to "integer" or "int32" and not waffle between them.

marstr replied:
Good question, the SDKs that get generated at times only support the most recent API Version. For that reason, there's some value in leaving small troubles like this alone for the sake of preventing unforeseen problems. Someday, when all SDKs truly support Multiple API Versions, it will be easier to make changes, so long as they're not inside of a single API Version.

marstr reviewed (changes requested):
Before I get too much further, would you mind fixing up the Semantic errors that were caught during CI?
 You can find that information here: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/290090374#L692

Commit:
Add back int format.

marstr commented:
@johanste / @fearthecowboy would you mind taking a look at the new functionality that is added in this PR too?

marstr commented:
Also, for folks reviewing, I've rebase this on my fork to help ignore the whitespace changes that were made as part of this PR. You can see my version here: https://github.com/marstr/azure-rest-api-specs/tree/FixFaceSwagger

davidwa-msft commented:
The whitespace changes in my change are desired. They came out of a swagger generation tool (Swashbuckle) and not hand-edited. The pre-existing swagger had many one-off edits following generation as well as validation errors that I fixed by making the desired swagger spec directly generated by the tool (for example, the missing int32 format issue was because it was one-off hand-edit, and I fixed it by making it directly associated and generated by code).

marstr commented:
yep! for sure @davidwa-msft! We'll make sure to preserve them, but for the sake of reviewing I'd like to isolate them into one commit so that we can consider just the semantic changes that you made. :)

fearthecowboy commented on specification/advisor/resource-manager/Microsoft.Advisor/2017-04-19/swagger/advisor.json:
```diff
@@ -6,18 +6,277 @@
 		"description": "REST APIs for Azure Advisor"
 	},
```
No, this spacing is not ok.

If you are generating this, you have to change the output or put it thru a filter and get the indentation down to 2 spaces.

davidwa-msft replied:
Fixed with latest change.

Commit:
Indent swagger with 2 spaces.
Remove unnecessary PATCH verb.
Fix ConfigData type in schema.
Fix examples to match.

davidwa-msft commented:
Are there any more concerns to address?

We have updated our swagger to use two-space indents, and we also removed the PATCH verb that we added in this change as we realized it was no longer necessary, so the change is smaller.

fearthecowboy commented:
Ouch, please don't move the file -- 

`specification/advisor/resource-manager/Microsoft.Advisor/2017-04-19/swagger/advisor.json` 

should be 

`specification/advisor/resource-manager/Microsoft.Advisor/2017-04-19/advisor.json

I'll finish reviewing this today.

davidwa-msft commented:
I was reading docs and other team's structure and realized that the swagger and the examples should be sibling directories instead of parent/child. So, I first made a commit to move it, then follow up diffs.

Commit:
Move swagger back.
Update x-ms-example path to match.
Update example name to Put prefix to match verb change following ARM review.

davidwa-msft commented:
I moved the swagger back to the original location and updated the x-ms-examples links to match and pass validation.

marstr reviewed (approved):
Chatted with @fearthecowboy. We wouldn't mind adding more client-side validation, but IMO that can wait for a future PR. I'll get this merged in once the CI jobs I've just restarted pass.

edit: finishing my thought

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-python

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-node

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-ruby
